src/fsharp/service/ServiceAnalysis.fs

```diff
@@ -153,45 +155,65 @@ module UnusedOpens =
         let newlyUsedOpenedGroups = 
             openedGroupsToExamine |> List.filter (fun openedGroup ->
                 openedGroup.OpenedModules |> List.exists (fun openedEntity ->
-
-                    symbolUses1 |> Array.exists (fun symbolUse ->
-                        rangeContainsRange openStatement.AppliedScope symbolUse.RangeAlternate &&
-                        match symbolUse.Symbol with
-                        | :? FSharpMemberOrFunctionOrValue as f ->
-                            match f.DeclaringEntity with
-                            | Some ent when ent.IsNamespace || ent.IsFSharpModule -> ent.IsEffectivelySameAs openedEntity.Entity
-                            | _ -> false
-                        | _ -> false) || 
+                    (match symbolUsesRangesByDeclaringEntity.TryGetValue openedEntity.Entity with
+                     | true, symbolUsesRanges ->
+                        symbolUsesRanges |> List.exists (fun symbolUseRange ->
+                            rangeContainsRange openStatement.AppliedScope symbolUseRange &&
+                            Range.posGt symbolUseRange.Start openStatement.Range.End)
+                     | _ -> false
+                    ) || 
 
                     symbolUses2 |> Array.exists (fun symbolUse ->
                         rangeContainsRange openStatement.AppliedScope symbolUse.RangeAlternate &&
+                        Range.posGt symbolUse.RangeAlternate.Start openStatement.Range.End &&
                         openedEntity.RevealedSymbolsContains symbolUse.Symbol)))
 
         // Return them as interim used entities
-        newlyUsedOpenedGroups |> List.collect (fun openedGroup -> 
-                openedGroup.OpenedModules |> List.map (fun x -> { Module = x.Entity; AppliedScope = openStatement.AppliedScope }))
+        let newlyOpenedModules = newlyUsedOpenedGroups |> List.collect (fun openedGroup -> openedGroup.OpenedModules)
+        for openedModule in newlyOpenedModules do
+            let scopes =
+                match usedModules.TryGetValue openedModule.Entity with
```

Comment:
Feeling like `dict.BagAdd(key, value)` and `dict.BagExistsValueForKey(key, f)` extension members in `illib.fs` might bring increased code clarity. 

They would need to be a C#-style extension member since they only operate on the type specific `Dictionary<'Key, 'Value list>`

Reply:
If done for clarity purpose, it might make sense to have a higher level data structure for the purpose of unused open analyzer.

Even if we add extension methods, we are still manipulating a raw dictionary directly in the code while a specific construct would hide that implementation specific detail and encapsulate the higher level operations behind more conveying names.

Albeit this is more work than just adding extension methods.